[Bug] Add Expr OverflowInTableInsert to fix AnsiCastOpSuite

amahussein opened the pull request:
Signed-off-by: Ahmed Hussein (amahussein) <a@ahussein.me>

fixes #6165

**Root Problem:**
The bug was caused by https://github.com/NVIDIA/spark-rapids/issues/6185

**Solution:**

- Revisited GpuCast to throw `ArithmeticOverflowException` when the overflow is detected. Before that, GpuCast could throw illegalStateException which does not match Spark behavior.
- Added shims for 331+ with:
  - new expression `GpuCheckOverflowInTableInsert` that delegates to the child expression `GpuCast`
  - new method added to throw  the Spark exception 
- Modified the AnsiCastOpSuite to handle the plan generated to get the correct count of the AnsiCast operations
- Modified the AnsiCastOpSuite to check for the exception messages expected during overflow in  a table insert.
- Modified the CastOpSuite to check for the right exception message.
- No modifications to integration tests

**Issue created as a followup:**

- Although this PR changed the exceptions of GpuCast to ArithmeticException, we may need a followup to make the exception message closer to the one generated in Spark. I will create a new issue to keep track of the list of issues related to ErrorMessages.

amahussein commented:
build

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCheckOverflowInTableInsert.scala:
```diff
@@ -0,0 +1,70 @@
+    // There are three different error messages: INVALID_INPUT, OVERFLOW, and INVALID_NUMBER.
+    // Currently, we treat Arithmetic and IllegalState as an overflow.
+    e match {
+      case _: ArithmeticException | _: IllegalStateException =>
+        child.isInstanceOf[GpuCast] &&
+          (e.getMessage.contains(GpuCast.INVALID_INPUT_MESSAGE)
+            || e.getMessage.contains(GpuCast.OVERFLOW_MESSAGE))
```
This is quite hacky and fragile.  When code resorts to scraping exception messages, something's gone wrong.  #5196 was supposed to update the plugin to throw SparkArithmeticException, and ideally we should just use the same exception and check for that one exception type here as Spark does. If we're not prepared to do that, there needs to be a followup as doing this type of logic is very problematic to maintain in practice.

amahussein replied:
There is an issue #6185 that I wanted to keep as followup.
The reason I did not go over changing `GpuCast` is that auditing Spark-3.3 and spark-3.4 show that there are still frequent updates in the error classes: #6182 , #6185, #6076, and #6262.
So, I thought that changing GpuCast can go out of scope of getting AnsiCastSuite to work on `[_,3.x[`. Beside 3.4 is not building.

jlowe commented on tests/src/test/scala/com/nvidia/spark/rapids/CastOpSuite.scala:
```diff
@@ -1123,6 +1123,8 @@ class CastOpSuite extends GpuExpressionTestSuite {
 
 /** Data shared between CastOpSuite and AnsiCastOpSuite. */
 object CastOpSuite {
+  val SPARK_OVERFLOW_INSERT_CLASS_NAME =
+    "org.apache.spark.sql.catalyst.expressions.CheckOverflowInTableInsert"
```
Why is this declared here when it's used in AnsiCastOpSuite?  Why not put it where it's needed?

revans2 commented on sql-plugin/src/main/311until320-nondb/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:
```diff
@@ -70,4 +70,12 @@ object RapidsErrorUtils {
+      from: DataType,
+      to: DataType,
+      columnName: String): ArithmeticException = {
+    // We should never call this method because GpuCheckOverflowInTableInsert cannot be in hthe plan
```
nit: hthe should be the

revans2 commented on sql-plugin/src/main/31xdb/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:
```diff
@@ -70,4 +70,12 @@ object RapidsErrorUtils {
+      from: DataType,
+      to: DataType,
+      columnName: String): ArithmeticException = {
+    // We should never call this method because GpuCheckOverflowInTableInsert cannot be in hthe plan
```
nit: same here hthe => the

revans2 commented on sql-plugin/src/main/320until330-nondb/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:
```diff
@@ -72,4 +72,12 @@ object RapidsErrorUtils {
+      from: DataType,
+      to: DataType,
+      columnName: String): ArithmeticException = {
+    // We should never call this method because GpuCheckOverflowInTableInsert cannot be in hthe plan
```
and here

revans2 commented on sql-plugin/src/main/321db/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:
```diff
@@ -75,4 +75,12 @@ object RapidsErrorUtils {
+      from: DataType,
+      to: DataType,
+      columnName: String): ArithmeticException = {
+    // We should never call this method because GpuCheckOverflowInTableInsert cannot be in hthe plan
```
and here

revans2 reviewed (commented):
Mostly have the same questions/concerns as Jason.

amahussein commented on sql-plugin/src/main/331+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:
```diff
@@ -0,0 +1,87 @@
+      requiredFieldName, matchedFields)
+  }
+
+  // To support Spark-3.3.1+ CAST_OVERFLOW_IN_TABLE_INSERT
```
I see. I thought that doing that way will be helpful if the changes are pulled into older branches.

amahussein commented:
build

amahussein commented:
Thanks @revans2 and @jlowe for the suggestions.
I updated the PR addressing your concerns:

- I filed a new issue #6293 as a place holder for issues to be addressed with error handling.
- I updated the PR description to reflect the current solution.
- I pushed the new commit to implement the fix as suggested:
  - Moved `GpuCheckOverflowInTableInsert` to the Shims
  - Changed the exception thrown by the `GpuCast` from `IllegalStateException` to `SparkArithmeticException`

revans2 commented on sql-plugin/src/main/331+/scala/com/nvidia/spark/rapids/GpuCheckOverflowInTableInsert.scala:
```diff
@@ -0,0 +1,55 @@
+      child.columnarEval(batch)
+    } catch {
+      // Convert ArithmeticException to Spark error class "CAST_OVERFLOW_IN_TABLE_INSERT"
+      case _: ArithmeticException =>
```
Why is the ArithmeticException and not SparkArithmeticException?

https://github.com/apache/spark/blob/7ae2b85c485a2f7ceb32543c64f08f3b1f120464/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L2371

amahussein replied:
Changed to `SparkArithmeticException`.
Initially I opted to `ArithmeticException` because this is the way it is used across the plugin code and the flexibility of using the code outside spark package.

revans2 commented on sql-plugin/src/main/331+/scala/com/nvidia/spark/rapids/GpuCheckOverflowInTableInsert.scala:
```diff
@@ -0,0 +1,55 @@
+ * QueryExecutionError.
+ * The calculation of the sideEffect is delegated to the child Expression "GpuCast".
+ */
+case class GpuCheckOverflowInTableInsert(child: Expression, columnName: String)
```
Why is child an Expression and not GpuCast like Spark enforces?

https://github.com/apache/spark/blob/7ae2b85c485a2f7ceb32543c64f08f3b1f120464/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L2364

amahussein replied:
I wish if Spark uses `Expression` because Spark-3.4 has different classes for Case.
Anyway, I changed that argument from `Expression` to `GpuCast`, but I thought that  `Expression` would be better to avoid failures when the "Replacement" changes in the future.

revans2 commented on sql-plugin/src/main/331+/scala/com/nvidia/spark/rapids/GpuCheckOverflowInTableInsert.scala:
```diff
@@ -0,0 +1,55 @@
+    }
+  }
+
+  override def toString: String = s"GpuCheckOverflowInTableInsert($child, $columnName)"
```
nit: The Spark version also has this be just the child.toString. It hides it from the end user, along with .sql too.

https://github.com/apache/spark/blob/7ae2b85c485a2f7ceb32543c64f08f3b1f120464/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L2404

amahussein replied:
yes, this was added as followup in Spark-3.4
For Spark-3.3 toString, and sql are not implemented . It appears in `plan.explain(true)`
https://github.com/apache/spark/blob/branch-3.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L2360  
Should we keep it that way for 3.3 and hide it for 3.4?

revans2 replied:
This is a place where I think it would be better to just follow 3.4 and not worry about it for 3.3. It is too complicated to try and make it match perfectly and it appears that they treated it as a "fix" Unless I am missing something.

amahussein replied:
I see! I changed the code to match Spark-3.4.

amahussein commented:
build

jlowe commented on sql-plugin/src/main/331+/scala/com/nvidia/spark/rapids/GpuCheckOverflowInTableInsert.scala:
```diff
@@ -0,0 +1,55 @@
+    try {
+      child.columnarEval(batch)
+    } catch {
+      // map SparkArithmeticException to SparkArithmeticException("CAST_OVERFLOW_IN_TABLE_INSERT")
+      case e if CastOverflowUtil.isCastOverflowException(e) =>
+        throw CastOverflowUtil.castCausingOverflowInTableInsertError(
+          child.child.dataType, child.dataType, columnName)
```
CastOverflowUtil seems like an abstraction that isn't worth the boilerplate.  It's called in exactly one place (here), and there's even a comment in CastOverflowUtil stating it's for here.  Things get a lot simpler to read and the code more closely matches the Spark implementation if we don't overengineer this until we actually need the abstractions.  YAGNI applies here, IMO.
```suggestion
      case e SparkArithmeticException =>
        throw QueryExecutionErrors.castingCauseOverflowErrorInTableInsert(
          child.child.dataType,
          child.dataType,
          columnName)
```

amahussein replied:
`QueryExecutionErrors` and `SparkArithmeticException` are both private to `Spark`. So, I followed the usual shim pattern of abstracting Spark private classes/methods.

jlowe replied:
The simpler approach is to put GpuCheckOverflowInTableInsert into the same package CastOverflowUtil.  It's overall less code to do so given how trivial GpuCheckOverflowInTableInsert is.

amahussein commented:
build
